rust_icu_umsg/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+    line: 0,
+    offset: 0,
+    preContext: [0; 16usize],
+    postContext: [0; 16usize],
+};
+
+/// Converts a parse error to a Result.
+fn parse_ok(e: sys::UParseError) -> Result<(), common::Error> {
+    if e == NO_PARSE_ERROR {
+        return Ok(());
+    }
+    Err(common::Error::Wrapper(anyhow!(
+        "parse error: line: {}, offset: {}",
+        e.line,
+        e.offset
+    )))
+}
+
+/// The implementation of the ICU `UMessageFormat*`.
+///
+/// Implements `UMessageFormat`
+#[derive(Debug)]
+pub struct UMessageFormat {
+    rep: *mut sys::UMessageFormat,
+}
+
+impl Drop for UMessageFormat {
+    /// Implements `umsg_close`.
+    fn drop(&mut self) {
+        unsafe {
+            versioned_function!(umsg_close)(self.rep);
+        }
+    }
+}
+
+/// The formatting arguments to be used in the call to `UMessageFormat::format`.
+///
+/// According to the [ICU documentation for `umsg_format`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/umsg_8h.html#a90a4b5fe778754e5da52f7c2e5fd6048):
+///
+/// > for all numeric arguments double is assumed unless the type is explicitly
+/// > integer (long).  All choice format arguments must be of type double.
+///
+/// Additionally, we determined by inspection that the string format is necessarily
+/// `rust_icu_ustring::UChar`.
+pub enum FmtArg {
+    /// A formatting argument that is a floating-point number.
+    Double(f64),
+    /// A formatting argument that is an integer.
+    Int(i32),
+    /// A formatting argument that is a string.
+    String(ustring::UChar),
+    /// A formatting argument that is a long integer.
+    Long(i64),
+
+    CStr(String),
+}
+
+impl UMessageFormat {
+    /// Creates a new message formatter.
+    ///
+    /// A single message formatter is created per each pattern-locale combination. Mutable
+    /// methods from `umsg` are not implemented, and for now requires that all formatting be
+    /// separate.
+    ///
+    /// Implements `umsg_open`.
+    pub fn try_from(
+        pattern: &ustring::UChar,
+        locale: &uloc::ULoc,
+    ) -> Result<UMessageFormat, common::Error> {
+        let pstr = pattern.as_c_ptr();
+        let loc = locale.as_c_str();
+        let mut status = common::Error::OK_CODE;
+        let mut parse_status = NO_PARSE_ERROR.clone();
+
+        let rep = unsafe {
+            assert!(common::Error::is_ok(status));
+            versioned_function!(umsg_open)(
+                pstr,
+                pattern.len() as i32,
+                loc.as_ptr(),
+                &mut parse_status,
+                &mut status,
+            )
+        };
+        common::Error::ok_or_warning(status)?;
+        parse_ok(parse_status)?;
+        Ok(UMessageFormat { rep })
+    }
+
+    /// Formats a new message based on this formatter.
+    ///
+    /// Implements `umsg_format`.
+    /// Implements `umsg_vformat`.
+    pub fn format(&self, args: &[FmtArg]) -> Result<String, common::Error> {
```

Comment:
`args` should be `&[impl Into<FmtArg>]` or something similar.

Reply:
See above: not needed.